Hunting down a slow Rails request

mauro-oto opened the pull request:
Hey @etagwerker, @schmierkov, 

I wrote a blog post on how I fixed the problem with a slow Rails API request.

I am not sure whether we should use the `EmailAccount` and `EmailAccountSync` model names, or we should change that to be something else, just for privacy/secrecy/legal matters. Maybe we could ask Preston for permission first. What do you think? 

Please check it out, thanks!

etagwerker commented on _posts/2016-03-04-hunting-down-a-slow-rails-request.markdown:
```diff
@@ -0,0 +1,97 @@
+author: "mauro-oto"
+---
+
+We have recently began using [Skylight](https://www.skylight.io) in production
```
@mauro-oto I think it'd sound more natural if you said: "Recently we started using... "

etagwerker commented on _posts/2016-03-04-hunting-down-a-slow-rails-request.markdown:
```diff
@@ -0,0 +1,97 @@
+---
+
+We have recently began using [Skylight](https://www.skylight.io) in production
+for one of our client's Rails applications, as an attempt to try to improve the
```
"for one of our clients." sounds a bit better. 
I think that "in an attempt to" sounds bettern than "as an attempt"

etagwerker commented on _posts/2016-03-04-hunting-down-a-slow-rails-request.markdown:
```diff
@@ -0,0 +1,97 @@
+for one of our client's Rails applications, as an attempt to try to improve the
+performance of some of the more critical API endpoints.
+
+Skylight reports on both time taken, with a breakdown of time taken per SQL
```
@mauro-oto This sounds a bit weird, could you refactor it to something like this? 

Skylight reports: 

* The time each request takes to complete
* A breakdown of the time it takes per SQL query
* Object allocations per request

etagwerker commented on _posts/2016-03-04-hunting-down-a-slow-rails-request.markdown:
```diff
@@ -0,0 +1,97 @@
+
+![Skylight report](/blog/assets/images/high-object-allocation.png)
+
+This request would take anywhere from 400ms to 3000ms to respond, which is WAY,
```
@mauro-oto I'd just say "*WAY* too long"

etagwerker commented on _posts/2016-03-04-hunting-down-a-slow-rails-request.markdown:
````diff
@@ -0,0 +1,97 @@
+pp strings
+```
+
+As you can see, there are 700k+ allocated strings, and this increase in object
````
@mauro-oto I'd separate this into "As you can see, there are 700k+ allocated strings. It's very likely that this amount of allocations is the culprit for the slow performance."

etagwerker commented on _posts/2016-03-04-hunting-down-a-slow-rails-request.markdown:
```diff
@@ -0,0 +1,97 @@
+allocation is the likely culprit for the slow request.
+
+I dug a bit into the `EmailAccount` model and its associations, and noticed that
+this model has a one-to-many association with the `EmailAccountSync` model.
```
* "this model had a one-to-many association"

etagwerker commented on _posts/2016-03-04-hunting-down-a-slow-rails-request.markdown:
```diff
@@ -0,0 +1,97 @@
+
+I dug a bit into the `EmailAccount` model and its associations, and noticed that
+this model has a one-to-many association with the `EmailAccountSync` model.
+However, we were only returning one of model `EmailAccountSync`, supposedly, by
```
@mauro-oto I'd say "Supposedly we were only returning one `EmailAccountSync` record by using: "

etagwerker commented on _posts/2016-03-04-hunting-down-a-slow-rails-request.markdown:
````diff
@@ -0,0 +1,97 @@
+EmailAccountSync Load (21.9ms)  SELECT "email_account_syncs".* FROM "email_account_syncs"  WHERE "email_account_syncs"."email_account_id" IN (24, 23, 22, 21, 20, 19, 18, 17, 16, 15)  ORDER BY email_account_syncs.last_attempt_at desc
+```
+
+All `EmailAccountSync` instances for the loaded `EmailAccount` instances get
````
I'd change this to: 

"This meant that all `EmailAccountSync` instances were getting loaded for each `EmailAccount` record. The `limit` that should've been in the SQL statement was not there."

etagwerker commented:
@mauro-oto Just reviewed this PR. Great work! :beers:

mauro-oto commented:
@etagwerker just addressed your comments. 

I also made `ul`s the same font size as `p`s for the post excerpts, as the lists looked bigger than paragraphs in the blog index page, see the first post in http://www.ombulabs.com/blog. 

Please check it out, thanks!

etagwerker commented on _posts/2016-03-04-hunting-down-a-slow-rails-request.markdown:
````diff
@@ -0,0 +1,100 @@
+Completed 200 OK in 178ms (Views: 5.6ms | ActiveRecord: 2.3ms)
+```
+
+The object allocation is down by a huge amount, and the request now takes around
````
"the request now takes from 50s to 200ms to response. This is not yet completely efficient, but it is a much more acceptable performance."

mauro-oto commented:
@etagwerker thanks! Let me know if you have any other comments. :tada:

etagwerker commented:
@mauro-oto :+1: :tada:
